### DISCUSSION TITLE
Complex story-board like animations without storyboard?

### DISCUSSION THREAD

**Discussion by mixonomicon** [AvaloniaUI/Avalonia#11905] [Q&A]
Hi,

I see the storyboard has been abandoned in Avalonia in lieu of more CSS-like animations. That part is fine, really. toryboards are verbose and long winded. 

But I do have several migration windows which have really complex multi-element, multi-property concurrent animations, moving elements around, showin and hiding other, changing opacity, etc, etc. Before this was all defined in a single storyboard which was triggered at some point. For example like:

```
        <Storyboard x:Key="MainStory">
            <DoubleAnimation Storyboard.TargetName="SizingGrid" Storyboard.TargetProperty="(Grid.Width)" Duration="00:00:01" To="1200"/>
            <DoubleAnimation Storyboard.TargetName="SizingGrid" Storyboard.TargetProperty="(Grid.Height)" Duration="00:00:01" To="800"/>
            <DoubleAnimation Storyboard.TargetName="MainCanvas" Storyboard.TargetProperty="(Canvas.Opacity)" Duration="00:00:01" To="0.4"/>
            <DoubleAnimation Storyboard.TargetName="MainLogo" Storyboard.TargetProperty="(Canvas.Opacity)" Duration="00:00:01" To="0.8"/>
            <ThicknessAnimation Storyboard.TargetName="MainCanvas" Storyboard.TargetProperty="(Canvas.Margin)" Duration="00:00:01" To="0,0,14,14"/>
            <ThicknessAnimation Storyboard.TargetName="TitleBox" Storyboard.TargetProperty="(Canvas.Margin)" Duration="00:00:01" To="0,2,48,0"/>

            <BooleanAnimationUsingKeyFrames Storyboard.TargetName="LauncherWindow" Storyboard.TargetProperty="(Window.Topmost)">
                <DiscreteBooleanKeyFrame KeyTime="00:00:00" Value="False"/>
            </BooleanAnimationUsingKeyFrames>
            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="xxxButton" Storyboard.TargetProperty="(Button.Visibility)">
                <DiscreteObjectKeyFrame KeyTime="00:00:00" Value="{x:Static Visibility.Hidden}"/>
            </ObjectAnimationUsingKeyFrames>

            <DoubleAnimationUsingKeyFrames Storyboard.TargetName="BackgroundGrid" Storyboard.TargetProperty="(Grid.Opacity)">
                <DiscreteDoubleKeyFrame KeyTime="00:00:01" Value="1.0"/>
            </DoubleAnimationUsingKeyFrames>
            <DoubleAnimation Storyboard.TargetName="InitialBackgroundGrid" Storyboard.TargetProperty="(Grid.Opacity)" BeginTime="00:00:01" Duration="00:00:01" To="0.0"/>

            <!-- etc, etc, -->

        </Storyboard>
```

How is the similar thing achieved now in Avalonia? CSS-based approach seems to be targeted around animating specific properties of one specific element based on class assignment, to create some small scale transitional animations. How can you group all those changes into one, well, storyboard that can be started simultaneously? It feels kind of wrong to embed all these particular animations in elements themselves and then having to change the class of all of them, which is the only thing that currently pops to my mind. .

**Comment by maxkatz6** [comment]
While you are right, that storyboards, in general, are more verbose, your specific use case might be the opposite for the CSS-like animations.
This proposal possibly can make it easier: https://github.com/AvaloniaUI/Avalonia/issues/11537#issuecomment-1608539967

There are could of hints on how you can make it easier.

1. In 11.0-rc you don't necessarily need styles to trigger animation. You can define Animation instance in the Resources dictionary with an x:SetterTargetType on it. After that, you can read this animation from resources and run it. Not documented yet https://github.com/AvaloniaUI/Avalonia/pull/8572/files#diff-4e4d7bc41f3d6ce463c075eb9593239234f4f0736c980cf55aeab347713cd6abR16vv (not yet documented, but I am going to add it to coming 11.0 docs).
2. It seems you can replace some of your animations with transitions. I.e. add DoubleTransition on Opacity and Width/Height, ThicknessTransition on Margin. Booleans don't really need it. See https://docs.avaloniaui.net/docs/animations/transitions
3. Having transitions set, you can trigger them by using simpler styles, something like:
```xaml
<Window.Styles>
   <Style Selector="Window.mainStoryActivated">
       <Setter Property="Topmost" Value="False" />
       <Style Selector="Grid#SizingGrid">
           <Setter Property="Width" Value="1200" />
           <Setter Property="Height" Value="800" />
       </Style>
       <Style Selector="Canvas#MainCanvas, Image#MainLogo">
           <Setter Property="Opacity" Value="1" />
       </Style>
       <!-- and so on -->
   </Style>
</Window.Styles>
```
Note, I used nested styles here, which also is a feature of 11.0. In 0.10 you would need to make it "flat", but won't be much of a difference here.

The obvious disadvantage is that your transitions will be set on controls always. Causing inconvenience if you change these properties outside of this main animation.

**Reply by mixonomicon** [reply]
I am sort of stuck on the step 1. 

I tried creating an animation resource, but don't seem to be able to target anything from it. This below does not work, as there is no target on the Setter, and omitting is probably just applying this to the Window. The `Animation` seems to only accept other animations and `KeyFrame` elements. And key frames only accept setters. 

``` xaml
    <Window.Resources>
        <Animation x:Name="MainStory" Duration="0:0:1" x:SetterTargetType="{x:Type Window}">
            <KeyFrame Cue="100%">
                <Setter ***Target="SizingGrid"*** Property="Width" Value="1200"/>
            </KeyFrame>
        </Animation>

```

**Reply by maxkatz6** [reply]
SetterTargetType defines type that your setter targets. If it's window, then setters should use window properties as setters. And then this animation can be started with `((Animation)Resources["MainStory"]).RunAsync(myWindow)`
Also, not "Name", but "Key", when it's in resources.

Note, these are not steps, but just hints, that you can use while porting.
Your specific use case probably is easier solvable with transitions.

**Reply by mixonomicon** [reply]
> SetterTargetType defines type that your setter targets. If it's window, then setters should use window properties as setters.

Ah, ok, now I get it. I thought this was just a starting type, whose logical sub-tree you can then access from embedded KeyFrames, but you can actually only animate one object, same as with CSS approach. I like your comment on #11537, this is actually what would have solved the issue (though this is realistically returning to a storyboard 😄  ) 
However, for now, it doesn't seem to really apply, as I would have to create a `SetterTargetType` resource for every element separately and then have a long list of `RunAsync()` calls to kick them all (potentially causing a timeshift between them?). 

I'll change my approach and try to complete my task with transitions.

**Reply by mixonomicon** [reply]
Just an additional info, for v11 and nested styles, the nested styles themselves seem to require nested selector (^), which is also undocumented. I tried appllying the above transitions but I got the "Child styles must have a nesting selector." error originating in `Avalonia.Styling.Selector.ValidateNestingSelector`. 

When I changed to below, it worked: 
``` xaml
<Window.Styles>
   <Style Selector="Window.mainStoryActivated">
       <Setter Property="Topmost" Value="False" />
       <Style Selector="^ Grid#SizingGrid">
...
```

notice the ^ in "^ Grid#SizingGrid"

in case anyone runs into this before the documentation is updated.

**Reply by mixonomicon** [reply]
@maxkatz6 
Hi, ok, I managed to do almost everything using steps 2 and 3 and nested styles. I just had to move animated properties into a default style, because transition/style priority is lower than the set values. For now I kicked off the transition using Avalonia Xaml Behavior project and setting an `.animated` style on the Window using `AddClassAction`, but thats not really relevant for this discussion.

One remaining issue I have is the "animation" or "transition", if you like, of a discrete value. 

Namely, after the overall transition is over, which lasts 2 seconds, I want to switch `SystemDecorations` on the Window to `Full` (it is `None` initially), to restore the chrome. If I do it immediately in Setter it works, but its ugly because it happens at moment 0:0:0 instead of at 0:0:2. If I try to transition it or animate it, it wont work because there does not seem to be a discrete (object) transitioner (compile error), and consequently no animator that can be used in key frames (blows up in runtime).

What can I do to set a discrete value at moment T > 0 after some transition has been activated?
